cpp/contracts/adapter.cpp

```diff
@@ -147,28 +152,104 @@ void adapter::freeuserdata(const name& account) {
+   require_auth(get_self());
+   check(topic_zero.size() == 32, "Expected 32 bytes emitter");
+   check(chain_id.size() == 32, "Expected 32 bytes chain_id");
+   mappings_table _mappings_table(get_self(), get_self().value);
+
+   auto mappings_itr = _mappings_table.find(pam::get_mappings_key(chain_id));
+
+   if (mappings_itr == _mappings_table.end()) {
+      _mappings_table.emplace(get_self(), [&](auto& row) {
+         row.chain_id = chain_id;
+         row.topic_zero = topic_zero;
+      });
+
+      print("Added a new mapping for chain_id: ", pam::get_mappings_key(chain_id));
+   } else {
+      _mappings_table.modify(mappings_itr, get_self(), [&](auto& row) {
+         row.topic_zero = topic_zero;
+      });
+
+      print("Updated the topic zero for chain_id: ", pam::get_mappings_key(chain_id));
+   }
+}
+
+void adapter::setemitter(bytes chain_id , bytes emitter) {
+   print("set emitter");
+   require_auth(get_self());
+   check(emitter.size() == 32, "Expected 32 bytes emitter");
+   check(chain_id.size() == 32, "Expected 32 bytes chain_id");
+   mappings_table _mappings_table(get_self(), get_self().value);
+
+   auto mappings_itr = _mappings_table.find(pam::get_mappings_key(chain_id));
+   if (mappings_itr == _mappings_table.end()) {
+      _mappings_table.emplace(get_self(), [&](auto& row) {
+         row.chain_id = chain_id;
+         row.emitter = emitter;
+      });
+
+      print("Added a new mapping for chain_id: ", pam::get_mappings_key(chain_id));
+   } else {
+      _mappings_table.modify(mappings_itr, get_self(), [&](auto& row) {
+         row.emitter = emitter;
+      });
+
+      print("Updated the emitter for chain_id: ", pam::get_mappings_key(chain_id));
+   }
+}
+
 void adapter::settle(const name& caller, const operation& operation, const metadata& metadata) {
    require_auth(caller);
 
    registry_adapter _registry(get_self(), get_self().value);
    auto idx_registry = _registry.get_index<adapter_registry_idx_token_bytes>();
    auto search_token_bytes = idx_registry.find(operation.token);
-
    check(search_token_bytes != idx_registry.end(), "invalid token");
 
-   checksum256 event_id;
-   check(pam::is_authorized(operation, metadata, event_id), "unauthorized");
+   checksum256 event_id = sha256((const char*)metadata.preimage.data(), metadata.preimage.size());
+   tee_pubkey _tee_pubkey(get_self(), get_self().value);
+   public_key tee_key = _tee_pubkey.get().key;
+
+   uint128_t a = 2;
+   bytes origin_chain_id = pam::extract_32bytes(metadata.preimage, a);
+
+   mappings_table _mappings_table(get_self(), get_self().value);
+
+   auto itr_mappings = _mappings_table.find(pam::get_mappings_key(origin_chain_id));
+   check(itr_mappings != _mappings_table.end(), "Unauthorized: origin chain_id not registered");
+   bytes exp_emitter = itr_mappings->emitter;
+   bytes exp_topic_zero =  itr_mappings->topic_zero;
+   pam::check_authorization(operation, metadata, event_id, tee_key, exp_emitter, exp_topic_zero);
```

Comment:
Yeah I would push to bring everything into the `is_authorized` method if possible